File: Examples/sample.py
"""
Prism Python sample.

Exercises:
- single, double, triple-quoted strings
- raw / byte / f-string prefixes
- numeric literals: int, float, hex, oct, bin, complex
- decorators
- class / def colorization
- async / await keywords
"""

from __future__ import annotations

import asyncio
import math
from dataclasses import dataclass
from typing import Iterable, Optional


# Numeric showcase
PI: float = 3.14159265358979
HEX_MASK: int = 0xDEAD_BEEF
OCTAL_PERMS: int = 0o755
BINARY_FLAG: int = 0b1010_1010
COMPLEX_NUMBER: complex = 3 + 4j


def fizzbuzz(n: int) -> Iterable[str]:
    r"""Classic fizzbuzz with a raw-string regex tribute: r"\d+"."""
    for i in range(1, n + 1):
        if i % 15 == 0:
            yield "FizzBuzz"
        elif i % 3 == 0:
            yield "Fizz"
        elif i % 5 == 0:
            yield "Buzz"
        else:
            yield str(i)


@dataclass(frozen=True)
class Point:
    x: float
    y: float

    def distance_to(self, other: "Point") -> float:
        return math.hypot(self.x - other.x, self.y - other.y)


class Greeter:
    """A trivial class to demonstrate `class` highlighting."""

    DEFAULT_GREETING: str = "Merhaba"

    def __init__(self, name: str, greeting: Optional[str] = None) -> None:
        self.name = name
        self.greeting = greeting or self.DEFAULT_GREETING

    def __call__(self) -> str:
        # f-string interpolation: keep the prefix highlighted.
        return f"{self.greeting}, {self.name}!"


async def fetch_squares(values: list[int]) -> list[int]:
    """Pretend to fetch squares asynchronously."""
    await asyncio.sleep(0)
    return [v ** 2 for v in values]


def main() -> None:
    bytes_literal = b"raw\xdebytes"
    rb_string = rb"verbatim\nbytes"  # noqa: F841
    print(f"hex={HEX_MASK:#x}, oct={OCTAL_PERMS:#o}, bin={BINARY_FLAG:#b}")
    print(", ".join(fizzbuzz(15)))

    g = Greeter("World")
    print(g())

    squares = asyncio.run(fetch_squares([1, 2, 3, 4]))
    assert squares == [1, 4, 9, 16], "math is broken"


if __name__ == "__main__":
    main()

File: Examples/sample.swift
// Prism Swift sample.
//
// Exercises:
// - line + block comments (block /* nested /* like this */ */ )
// - regular, multi-line, and raw strings (#"..."#, ##"..."##)
// - numeric literals: int, double, hex, octal, binary
// - attributes (@available, @MainActor, @propertyWrapper)
// - struct / class / enum / protocol / extension / actor
// - async / await / throws / try

import Foundation

// MARK: - Constants

let pi: Double = 3.14159_26535
let mask: UInt32 = 0xDEAD_BEEF
let perms: Int = 0o755
let flag: UInt8 = 0b1010_1010
let scientific: Double = 6.022e23

// Strings — every flavor.
let plain = "hello, world"
let interpolated = "pi ≈ \(pi)"
let multiline = """
    This is a
    multi-line string
    with \(interpolated) interpolation.
    """
let rawSingle = #"a regex with literal backslash: \d+"#
let rawTriple = ##"contains a "# without closing"##

// MARK: - Protocols

protocol Greetable {
    var name: String { get }
    func greet() -> String
}

// MARK: - Types

struct Point: Equatable {
    let x: Double
    let y: Double

    func distance(to other: Point) -> Double {
        let dx = x - other.x
        let dy = y - other.y
        return (dx * dx + dy * dy).squareRoot()
    }
}

enum Direction: String, CaseIterable {
    case north, south, east, west

    var opposite: Direction {
        switch self {
        case .north: return .south
        case .south: return .north
        case .east:  return .west
        case .west:  return .east
        }
    }
}

actor Counter {
    private var value: Int = 0

    func increment() { value += 1 }
    func read() -> Int { value }
}

@MainActor
final class Greeter: Greetable {
    let name: String

    init(name: String) {
        self.name = name
    }

    func greet() -> String {
        "Merhaba, \(name)!"
    }
}

// MARK: - Property wrapper

@propertyWrapper
struct Clamped<Value: Comparable> {
    private var storage: Value
    private let range: ClosedRange<Value>

    init(wrappedValue: Value, _ range: ClosedRange<Value>) {
        self.range = range
        self.storage = min(max(wrappedValue, range.lowerBound), range.upperBound)
    }

    var wrappedValue: Value {
        get { storage }
        set { storage = min(max(newValue, range.lowerBound), range.upperBound) }
    }
}

// MARK: - Async showcase

@available(macOS 12.0, *)
func fetchSquares(_ values: [Int]) async throws -> [Int] {
    try await Task.sleep(nanoseconds: 1)
    return values.map { $0 * $0 }
}

// MARK: - Extension

extension Array where Element == Int {
    /// Sum of squares — generic-constrained extension.
    var sumOfSquares: Int { reduce(0) { $0 + $1 * $1 } }
}

#if DEBUG
let label = "debug build"
#else
let label = "release build"
#endif

print(label)
